Various minor fixes in preparation for hot-cert reload

mkrautz opened the pull request:
- Commit 1: Change MetaParams::typeCheckedFromSettings to take an optional QSettings parameter.
- Commit 2: Add USR1 signal handler implementation.
- Commit 3: Move callsite of qmConfig.clear() in MetaParams::read() to allow MetaParams::read() to be split out into separate functions.
- Commit 4: Add Server::bUsingMetaCert flag, such that hot cert reload knows which servers to reload.
- Commit 5: Add qAbsSettingsFn, such that hot cert reload knows where to read murmur.ini.
- Commit 6: Move SSL check and OpenSSL version advertisement from MetaParams::read() to main.cpp.
- Commit 7: Avoid use of global QSslSocket::defaultCiphers() list; prefer applying the default list to each QSslSocket/QSslConfiguration individually.

Kissaki commented on src/murmur/Meta.cpp:
```diff
@@ -458,49 +467,45 @@ void MetaParams::read(QString fname) {
+					 qVersion());
+			}
+
+			qlCiphers = filtered;
```
The entire codepath is broken (maybe from a conflict resolving?). `filtered` is defined twice, and the one that is never filled is used here.

mkrautz replied:
No, it is actually the correct one that is used. However, you are correct that there is a duplicate.
I moved the filtered var into the block when I refactored.

mkrautz commented on src/murmur/Meta.h:
```diff
@@ -125,6 +129,9 @@ class MetaParams {
 
+	/// qsAbsSettingsFn is the absolute path to
+	/// the murmur.ini used by this Meta instance.
+	QString qsAbsSettingsFn;
```
`fn` as in `filename`

Should I avoid the short-hand?

mkrautz replied:
Changed it to Filename.

Kissaki replied:
It’s a file path, not a file name tho, right? :)

The commit description still talks about `qAbsSettingsFn`

mkrautz commented on src/murmur/UnixMurmur.cpp:
```diff
@@ -149,6 +166,12 @@ void UnixMurmur::termSignalHandler(int) {
+	char a = 1;
+	ssize_t len = ::write(iUsr1Fd[0], &a, sizeof(a));
+	Q_UNUSED(len);
+}
```
It writes to iUser1Fd, which will then trigger a QSocketNotifier to read from the fd, which can then trigger the signal handling inside Qt's event loop.

It's basically a trick to move the signal handling into Qt, because signal contexts are very fragile.

mkrautz commented on src/murmur/UnixMurmur.cpp:
```diff
@@ -199,6 +222,17 @@ void UnixMurmur::handleSigTerm() {
+void UnixMurmur::handleSigUsr1() {
+	qsnUsr1->setEnabled(false);
+	char tmp;
+	ssize_t len = ::read(iUsr1Fd[1], &tmp, sizeof(tmp));
```
See previous explanation.

mkrautz commented on src/murmur/UnixMurmur.cpp:
```diff
@@ -121,20 +127,31 @@ UnixMurmur::UnixMurmur() {
 
+	usr1.sa_handler = usr1SignalHandler;
+	sigemptyset(&term.sa_mask);
+	term.sa_flags = SA_RESTART;
```
Yes

mkrautz commented on src/murmur/UnixMurmur.cpp:
```diff
@@ -121,20 +127,31 @@ UnixMurmur::UnixMurmur() {
 		qFatal("Failed to install SIGTERM handler");
 
+	usr1.sa_handler = usr1SignalHandler;
+	sigemptyset(&term.sa_mask);
```
Yes

mkrautz commented:
> A QSettings that is not the one known to Meta? Seems strange with Meta still holding a QSettings. Will have to see how it looks when used.

We need to reload the settings. The way that happens is that I create a new QSettings and update the MetaParams with the new SSL settings.

I don't even think MetaParam's own QSettings object is necessary once we've called MetaParams:.read(), will look into it...

> Commit 3 d5b773c has a logically duplicate paragraph it seems.

Fixed.

> Commit 4 4e3b6c0 says hot reloading will only be possible for meta certs. Why? Does it mean the new codepath that will be added via usr signal? Isn't hot loading already possible for vservers?

Kinda. When I say hot cert reloading, I mean the "send USR1 to the murmur process to reload the cert files specified in murmur.ini".

Should I be more specific?  The USR1 signal will not load certs for vservers, since they can't change without using RPC, and if you're using RPC, you should use the updateCertificate() server method instead.

> Commit 5 fdc87ad "in order to re-load the settings from the same file.".

Done

> Meta.cpp#L156 and L161 should be qsAbsSettingsFn rather than fname.

Fixed.

> We are only using the ciphers list in two places:
>
> New user connections
> Network requests to our registration service
> Shouldn't there also be our initial server listening?

Qt doesn't provide a SSL server socket, so the way we do it is we have a TCP server, and for each incoming connection, we wrap it in QSslSocket to "upgrade" it to a SSL connection.
This is how you're supposed to do it in Qt.
So: there isn't a listning SSL socket... They're all client sockets as far as the server is concerned.

mkrautz commented on src/murmur/Server.h:
```diff
@@ -131,6 +131,7 @@ class Server : public QThread {
 		QVariant qvSuggestPositional;
 		QVariant qvSuggestPushToTalk;
 
+		bool bUsingMetaCert;
```
Done

mkrautz commented:
PTAL.

New changes at https://github.com/mumble-voip/mumble/compare/master...mkrautz:hot-reload-certs-prep-v2

mkrautz commented:
Original patchset at https://github.com/mumble-voip/mumble/compare/master...mkrautz:hot-reload-certs-prep-v1

mkrautz commented:
This PR is now @ hot-reload-certs-prep-v2.

Kissaki commented:
> Should I be more specific?

Yes, I think so.

> we will only be able to perform our live reload for those
servers.

implies there is no live reloading for vservers, when it's just a different mechanism. Not sure how to formulate (point to separate mechanism, or word differently...).

Kissaki commented on src/murmur/Meta.cpp:
```diff
@@ -458,49 +467,45 @@ void MetaParams::read(QString fname) {
+			if (ciphers.size() != filtered.size()) {
+				qWarning("Warning: all cipher suites in sslCiphers using Diffie-Hellman key exchange "
+				         "have been removed. Qt %s does not support custom Diffie-Hellman parameters.",
+					 qVersion());
```
This is aligned with tabs and a space, but the above line with spaces. Should all be spaces?

Kissaki commented on src/murmur/Meta.cpp:
```diff
@@ -458,49 +467,44 @@ void MetaParams::read(QString fname) {
+			if (ciphers.size() != filtered.size()) {
+				qWarning("Warning: all cipher suites in sslCiphers using Diffie-Hellman key exchange "
+				         "have been removed. Qt %s does not support custom Diffie-Hellman parameters.",
+						 qVersion());
```
This is aligned with tabs and a space, but the above line with spaces. Should all be spaces?

mkrautz replied:
Done.

Kissaki commented:
Commit description still talks about `qAbsSettingsFn`

mkrautz commented:
@Kisski said:
>> Should I be more specific?
>
>Yes, I think so.
>
>> we will only be able to perform our live reload for those
servers.
>
> implies there is no live reloading for vservers, when it's just a different mechanism (that already can live reload/set). Not sure how to formulate (point to separate mechanism, or word differently...).

Rewrote the commit messages.

mkrautz commented:
> Commit description still talks about qAbsSettingsFn

Done. It's now called qAbsSettingsFilePath.

mkrautz commented:
The new changes are committed to this branch (hot-reload-certs-prep), but they're also available in https://github.com/mkrautz/mumble/tree/hot-reload-certs-prep-v3

v1: https://github.com/mkrautz/mumble/tree/hot-reload-certs-prep-v1 (**old**)
v2: https://github.com/mkrautz/mumble/tree/hot-reload-certs-prep-v2 (**previous**)
v3: https://github.com/mkrautz/mumble/tree/hot-reload-certs-prep-v3 (**current**)

mkrautz commented:
@Kissaki PTAL
